### PR TITLE
TDL-19647: Add missing tap-tester tests

### PR DESCRIPTION
**Pull request by harshpatel4crest**
# Description of change
TDL-19647: Add missing tap-tester tests
- Added missing tap-tester tests
  - Delimeters Test
  - Discovery Test
  - Start Date Test
  - Maximize CSV Field Width Test
  - Automatic Fields Test
  - All Fields Test

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

**Review by savan-chovatiya — CHANGES_REQUESTED** [review]

File: tests/test_start_date.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+        # Verify we synced more records in 2nd sync in comparision to 1st sync, as file is updated
+        self.assertGreater(len(actual_records_1), len(actual_records_2))
+
+        # Verify we get same bookmark for both syncs
+        self.assertEqual(
+            self.parse_date(state_2.get("bookmarks").get("employee_table").get("modified_since"), "%Y-%m-%dT%H:%M:%S+00:00"),
+            self.parse_date(state_1.get("bookmarks").get("employee_table").get("modified_since"), "%Y-%m-%dT%H:%M:%S+00:00"))
+
+        # verify we replicated sync 2 records in 1st sync too
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/test_start_date.py]: Updated comment.

File: tests/test_discovery.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            with self.subTest(stream=stream):
+
+                # Verify ensure the catalog is found for a given stream
+                catalog = next(iter([catalog for catalog in found_catalogs if catalog["stream_name"] == stream]))
+                self.assertIsNotNone(catalog)
+
+                # collecting expected values
+                expected_primary_keys = {'id'}
+
+                # collect actual values
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/test_discovery.py]: Updated the comment.

File: tests/test_discovery.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                        msg="One or more streams don't follow standard naming")
+
+        for stream in self.expected_check_streams():
+            with self.subTest(stream=stream):
+
+                # Verify ensure the catalog is found for a given stream
+                catalog = next(iter([catalog for catalog in found_catalogs if catalog["stream_name"] == stream]))
+                self.assertIsNotNone(catalog)
+
+                # collecting expected values
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/test_discovery.py]: Updated the comment.

File: tests/test_discovery.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        runner.run_check_job_and_check_status(self)
+
+        found_catalogs = menagerie.get_catalogs(self.conn_id)
+        self.assertEqual(len(found_catalogs), 1, msg="unable to locate schemas for connection {}".format(self.conn_id))
+
+        # Verify stream names follow naming convention
+        # streams should only have lowercase alphas and underscores
+        found_catalog_names = {c['tap_stream_id'] for c in found_catalogs}
+        self.assertTrue(all([re.fullmatch(r"[a-z_]+",  name) for name in found_catalog_names]),
+                        msg="One or more streams don't follow standard naming")
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/test_discovery.py]: Right. Commented the assertion and added the comment for the same.

File: tests/test_delimiters.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    def test_TXT_file_with_semi_colon_delimeter(self):
+        self.file_name = "semi_colon.txt"
+        self.delimeter = ";"
+        self.upload_file()
+        self.run_test()
+
+    def test_TXT_file_with_tab_delimeter(self):
+        self.file_name = "tab.txt"
+        self.delimeter = "\t"
+        self.upload_file()
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/test_delimiters.py]: Updated the code.

File: tests/test_delimiters.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+from utils_for_test import delete_and_push_file
+
+FOLDER_PATH = "Delimiters"
+
+class S3DelimetersBase(unittest.TestCase):
+
+    file_name = None
+
+    def name(self):
+        return "test_csv_file_with_with_delimeter"
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/test_delimiters.py]: Updated the test name.

File: tests/test_automatic_fields.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        self.run_and_verify_sync(self.conn_id)
+
+        synced_records = runner.get_records_from_target_output()
+
+        for stream in self.expected_sync_streams():
+            with self.subTest(stream=stream):
+
+                expected_keys = self.expected_automatic_fields().get(stream)
+
+                # collect actual values
```
Comment (dbshah1212) [inline tests/test_automatic_fields.py:57]:
```suggestion
                # Collect actual values
```

File: tests/test_automatic_fields.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+                data = synced_records.get(stream, {})
+                record_messages_keys = [set(row.get('data').keys()) for row in data.get('messages', {})]
+
+                for actual_keys in record_messages_keys:
+                    self.assertSetEqual(expected_keys, actual_keys)
+
+    def test_CSV_with_1_PK(self):
```
Comment (dbshah1212) [inline tests/test_automatic_fields.py:64]: Add code comments

File: tests/test_automatic_fields.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        self.file_name = "CSV_with_one_primary_key.csv"
+        self.headers = "id"
+        self.table_entry = [{"table_name": "csv_with_pk", "search_prefix": "tap_tester", "search_pattern": "{}".format(self.file_name), "key_properties": "{}".format(self.headers)}]
+        self.run_test()
+
+    def test_CSV_with_2_PK(self):
```
Comment (dbshah1212) [inline tests/test_automatic_fields.py:70]: Add code comments

File: tests/test_delimiters.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        return "test_csv_file_with_delimeter"
+
+    def resource_name(self):
+        return [self.file_name]
+
+    def upload_file(self):
+        delete_and_push_file(self.get_properties(), self.resource_name(), FOLDER_PATH)
+        self.conn_id = connections.ensure_connection(self)
+
+    def tap_name(self):
```
Comment (dbshah1212) [inline tests/test_delimiters.py:24]: Do we need this?

File: tests/test_delimiters.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+            {'name': 'Worthington', 'birthday': '1938-04-15T11:18:18Z', 'gender': 'Male', 'age': 42, 'bank_balance': Decimal('837.62'), 'mixed_nums': 206, '_sdc_source_bucket': 'com-stitchdata-prod-circleci-assets', '_sdc_source_file': 'tap_tester/Delimiters/{}'.format(self.file_name), '_sdc_source_lineno': 4},
+            {'name': 'Lorne', 'birthday': '1987-09-22T17:58:39Z', 'gender': 'Male', 'age': 10, 'bank_balance': Decimal('533.49'), 'mixed_nums': 137, '_sdc_source_bucket': 'com-stitchdata-prod-circleci-assets', '_sdc_source_file': 'tap_tester/Delimiters/{}'.format(self.file_name), '_sdc_source_lineno': 5},
+            {'name': 'Saundra', 'birthday': '2005-07-18T20:34:15Z', 'gender': 'Male', 'age': 76, 'bank_balance': Decimal('524.76'), 'mixed_nums': 252, '_sdc_source_bucket': 'com-stitchdata-prod-circleci-assets', '_sdc_source_file': 'tap_tester/Delimiters/{}'.format(self.file_name), '_sdc_source_lineno': 6}
+        ]
+
+        self.assertEqual(actual_records, expected_records)
+
+    def test_CSV_file_with_comma_delimeter(self):
```
Comment (dbshah1212) [inline tests/test_delimiters.py:82]: Add functional comments

File: tests/test_discovery.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    def resource_name(self):
+        return ["discovery_test.csv"]
+
+    def name(self):
+        return "test_start_date"
+
+    def expected_check_streams(self):
+        return {'employees'}
+
+    def get_credentials(self):
```
Comment (dbshah1212) [inline tests/test_discovery.py:33]: Do we need all these functions?

Comment (harshpatel4crest) [reply tests/test_discovery.py]: Removed the redundant function.

File: tests/test_delimiters.py (hunk 11 above)
Comment (harshpatel4crest) [reply tests/test_delimiters.py:73]: Added function comments.

File: tests/test_delimiters.py (hunk 10 above)
Comment (harshpatel4crest) [reply tests/test_delimiters.py]: Removed redundant function.

File: tests/test_automatic_fields.py (hunk 9 above)
Comment (harshpatel4crest) [reply tests/test_automatic_fields.py:73]: Added code comments.

File: tests/test_automatic_fields.py (hunk 8 above)
Comment (harshpatel4crest) [reply tests/test_automatic_fields.py:64]: Added code comments.

File: tests/test_automatic_fields.py (hunk 7 above)
Comment (harshpatel4crest) [reply tests/test_automatic_fields.py]: Updated the comment.

**Review by dbshah1212 — APPROVED** [review]

File: tests/test_maximize_csv_field_width.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+    def expected_pks(self):
+        return {'csv_with_max_field_width': {}}
+
+    def test_run(self):
+
+        found_catalogs = self.run_and_verify_check_mode(self.conn_id)
+
+        found_catalog_names = set(map(lambda c: c['tap_stream_id'], found_catalogs))
+        self.assertEqual(self.expected_check_streams(), found_catalog_names, msg="Expected check streams are not subset of discovered catalog")
```
Comment (kspeer825) [inline tests/test_maximize_csv_field_width.py:34]: I think this test should run a sync and validate the max width value is replicated

**Review by kspeer825 — CHANGES_REQUESTED** [review]
Left 1 suggestion

File: tests/test_maximize_csv_field_width.py (hunk 13 above)
Comment (harshpatel4crest) [reply tests/test_maximize_csv_field_width.py:34]: Updated the test case as per suggestion.

**Review by kspeer825 — APPROVED** [review]

**Pull request merged**